src/luaclang.c

```diff
@@ -5,50 +5,68 @@
 #include "lualib.h"
 #include "lauxlib.h"
 
-#define INDEX_METATABLE  "Clang.Index"
-#define TU_METATABLE     "Clang.TU"
+#define CLANG_METATABLE  "Clang.Object"
 #define CURSOR_METATABLE "Clang.Cursor"
 #define TYPE_METATABLE   "Clang.Type"
 
-/* Create CXIndex as userdata */
+typedef struct clang_object {
+        CXIndex idx;
+        CXTranslationUnit tu;
+} clang_object;
 
-static CXIndex *new_CXIndex(lua_State *L) 
+/* Create clang_object as userdata */
+static clang_object *new_clangobject(lua_State *L) 
 {
-        CXIndex *idx = (CXIndex*) lua_newuserdata(L, sizeof(CXIndex));
-        luaL_getmetatable(L, INDEX_METATABLE);
+        clang_object *clang_parser = (clang_object*) lua_newuserdata(L, sizeof(clang_object));
+        luaL_getmetatable(L, CLANG_METATABLE);
         lua_setmetatable(L, -2);
-        return idx;
+        return clang_parser;
 }
 
-/* Convert userdata type to CXIndex */
-
-static CXIndex to_CXIndex(lua_State *L, int n)
+/* Convert userdata type to  */
+static clang_object to_clangobject(lua_State *L, int n) 
```

Comment:
You should return a pointer, else you won't be able to assign NULL to its fields later.
Probably, this is the reason that you are getting a segfault.